### ISSUE TITLE
語音增強DCCRN專案執行計劃

### ISSUE THREAD

**Issue by Hank-Jiang40815** [Hank-Jiang40815/Knowledge_distillation#2]
## 專案執行計劃

本 Issue 提供了 DCCRN 語音增強模型專案的完整執行計劃，包括環境設置、數據準備、訓練、知識蒸餾和模型壓縮等步驟。

### 1. 環境設置
- [ ] 確認 Python 3.7.5 環境
- [ ] 安裝 PyTorch 1.10.2
- [ ] 安裝依賴庫：
```bash
pip install -r requirements.txt
```
- [ ] 確認 GPU 可用性和 CUDA 版本兼容性

### 2. 數據準備
- [ ] 確認數據集完整性 (dataset/clean 和 dataset/noisy)
- [ ] 如需要，執行數據解壓：
```bash
python dataset/unzip.py --config dataset/unzip_cfg.toml
```
- [ ] 生成訓練數據：
```bash
python dataset/gen.py --config dataset/gen_cfg.toml
```
- [ ] 確認 dataset_csv 目錄中的數據分割(train.csv, valid.csv, test.csv)

### 3. 基礎模型訓練
- [ ] 檢查並調整 config/base_config.toml 中的參數
- [ ] 創建訓練腳本 train.py，使用以下模式：
```python
from trainer import base_trainer

trainer = base_trainer.BaseTrainer(config_path="config/base_config.toml")
trainer.train()
```
- [ ] 執行訓練：
```bash
python train.py
```
- [ ] 監控 training loss 和 validation 指標
- [ ] 確認 checkpoint 保存在 output/dccrn/checkpoints 目錄

### 4. 模型評估
- [ ] 使用 compute_metrics.py 評估模型性能：
```bash
python dataset/compute_metrics.py --config dataset/compute_metrics_cfg.toml
```
- [ ] 記錄 PESQ, STOI 等指標結果

### 5. 知識蒸餾實現 (進階)
- [ ] 創建或修改 config/lite_v1_config.toml 用於輕量化模型
- [ ] 創建知識蒸餾訓練腳本 distill.py：
```python
from trainer import knowledge_distillation_trainer

trainer = knowledge_distillation_trainer.KnowledgeDistillationTrainer(
    teacher_model_path="backup/base_model.pth",
    student_config_path="config/lite_v1_config.toml"
)
trainer.train()
```
- [ ] 執行知識蒸餾：
```bash
python distill.py
```
- [ ] 比較蒸餾前後模型性能

### 6. 模型壓縮 (進階)
- [ ] 模型剪枝：
```python
from trainer import prune_trainer

trainer = prune_trainer.PruneTrainer(
    model_path="backup/lite_v1_model.pth",
    config_path="config/lite_v1_config.toml"
)
trainer.train()
```
- [ ] 模型量化：
```python
from trainer import quantization_trainer

trainer = quantization_trainer.QuantizationTrainer(
    model_path="backup/lite_v1_model.pth",
    config_path="config/lite_v1_config.toml"
)
trainer.train()
```
- [ ] 評估壓縮後模型的性能和大小

### 7. 推理測試
- [ ] 創建推理腳本 infer.py：
```python
from inferencer import base_inferencer

inferencer = base_inferencer.BaseInferencer(model_path="output/dccrn/checkpoints/best_model.pth")
enhanced_audio = inferencer.enhance("path/to/noisy_audio.wav")
inferencer.save_audio(enhanced_audio, "path/to/enhanced_audio.wav")
```
- [ ] 測試不同條件下的推理性能

### 重要注意事項
1. 確保不同操作使用一致的音頻參數 (sr=16000, n_fft=512, etc.)
2. 定期備份重要模型權重和實驗結果
3. 系統記錄實驗過程和結果，以便進行對比分析
4. 為接下來的工作做準備：實現 README 中提到的 DCCRN-C 和 DCCRN-C(light) 模型

### 參考資料
- DCCRN 論文：https://arxiv.org/abs/2008.00264
- 知識蒸餾論文：https://arxiv.org/abs/1503.02531
- 模型剪枝參考：https://pytorch.org/tutorials/intermediate/pruning_tutorial.html
- 模型量化參考：https://pytorch.org/docs/stable/quantization.html
